pkg/election/election.go

```diff
@@ -0,0 +1,284 @@
+		}
+		if err != nil && errors.Cause(err) != ctx.Err() { // only send non-ctx.Err() error
+			e.ech <- err
+		}
+	}()
+
+	for {
+		// check context canceled/timeout
+		select {
+		case <-session.Done():
+			e.l.Info("etcd session is done, will try to create a new one", zap.Int64("old lease", int64(session.Lease())))
+			closeSession(session)
+			session, err = e.newSession(ctx, newSessionRetryUnlimited) // retry until context is done
+			if err != nil {
+				err = terror.ErrElectionCampaignFail.Delegate(err, "create a new session")
+				return
+			}
+		case <-ctx.Done():
+			e.l.Info("break campaign loop, context is done", zap.Error(ctx.Err()))
+			return
+		default:
+		}
+
+		// try to campaign
+		elec := concurrency.NewElection(session, e.key)
+		err = elec.Campaign(ctx, e.id)
+		if err != nil {
+			// err may be ctx.Err(), but this can be handled in `case <-ctx.Done()`
+			e.l.Warn("fail to campaign", zap.Error(err))
+			continue
+		}
+
+		// compare with the current leader
+		leaderKey, leaderID, err := getLeaderInfo(ctx, elec)
+		if err != nil {
+			// err may be ctx.Err(), but this can be handled in `case <-ctx.Done()`
+			e.l.Warn("fail to get leader ID", zap.Error(err))
+			continue
+		}
+		if leaderID != e.id {
+			e.l.Info("current member is not the leader", zap.String("current member", e.id), zap.String("leader", leaderID))
+			continue
+		}
+
+		e.toBeLeader() // become the leader now
+		e.watchLeader(ctx, session, leaderKey)
+		e.retireLeader() // need to re-campaign
+	}
+}
+
+func (e *Election) toBeLeader() {
+	e.isLeader.Set(true)
+	select {
+	case e.leaderCh <- true:
+	default:
+	}
+	e.l.Info("become the leader", zap.String("member", e.id))
+}
+
+func (e *Election) retireLeader() {
+	e.isLeader.Set(false)
+	select {
+	case e.leaderCh <- false:
+	default:
+	}
+	e.l.Info("retire from the leader", zap.String("member", e.id))
+}
+
+func (e *Election) watchLeader(ctx context.Context, session *concurrency.Session, key string) {
+	e.l.Debug("watch leader key", zap.String("key", key))
+	wch := e.cli.Watch(ctx, key)
+	for {
+		select {
+		case resp, ok := <-wch:
+			if !ok {
+				e.l.Info("watch channel is closed")
+				return
+			}
+			if resp.Canceled {
+				e.l.Info("watch canceled")
+				return
+			}
+
+			for _, ev := range resp.Events {
+				if ev.Type == mvccpb.DELETE {
+					e.l.Info("fail to watch, the leader is deleted", zap.ByteString("key", ev.Kv.Key))
+					return
+				}
+			}
+		case <-session.Done():
+			return
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (e *Election) newSession(ctx context.Context, retryCnt int) (*concurrency.Session, error) {
+	var (
+		err     error
+		session *concurrency.Session
+	)
+	for i := 0; i < retryCnt; i++ {
+		if i > 0 {
+			select {
+			case e.ech <- terror.ErrElectionCampaignFail.Delegate(err, "create a new session"):
+			default:
+			}
+
+			select {
+			case <-time.After(newSessionRetryInterval):
+			case <-ctx.Done():
+				break
```

Comment:
do you means break the for loop, this break only break the `select`
https://stackoverflow.com/questions/11104085/in-go-does-a-break-statement-break-from-a-switch-select

Reply:
good catch!, fixed in https://github.com/pingcap/dm/pull/367/commits/b99ae4653c5039cfb5070e5b0d62d6bdad77599a.